### ISSUE TITLE
Berkshelf needs a way to run unattended (without a shell) without requiring $HOME to be set

### ISSUE THREAD

**Issue by eric-tucker** [berkshelf/berkshelf#1429]
Attempt 2.

When running commands unattended such as via cfn-init through CloudFormation or via AWS CodeDeploy, $HOME will not be set.

By default, Berks attempts to install cookbooks to $HOME/.berkshelf.  This obviously doesn't work with $HOME not set.

My thoughts were that the vendor command should be able to simply download cookbooks from the supermarket and place them in the specified directory, but it seems it needs to download first to $HOME/.berkshelf followed by copying them to the vendor directory.

It would be nice if Berks could run without requiring $HOME set in order to support unattended downloading of cookbooks.

**Comment by reset** [comment]
I've had some time to give this a thought and I think it's a perfectly reasonable request.

We currently cache everything in the user's home directory and assume the `$HOME` environment variable is set because Berkshelf was intended to be a dependency resolver for developers by means of their command line. I think it would be perfectly reasonable to change the location of where things are cached if the `$HOME` variable is not set or if we can detect that the process is not running as part of an interactive session. We use the `$BERKSHELF_HOME` environment variable right now to customize the location of Berkshelf's cache directory where cookbooks, configuration, and transient data is stored. However, it doesn't sound like that will solve your problem since you are using a non-interactive session.

Perhaps the best place to store the cache would be in the `/tmp` directory by default if we detect a non-interactive session or the `$HOME` env var is not set. We could also allow this location to be configured by the Berkshelf configuration or by adding a flag that is available to all Berkshelf commands.

This is outside of the normal use case for a lot of Berkshelf users, including the Berkshelf core team, so I don't think we'd add this in ourselves. If you'd like to submit a pull request or continue the discussion here I'd be more than happy to figure out how we can make this work for you. The gem is easy to contribute to, just clone the repository and use the standard rake commands for bundler to package the gem up after you've made your changes: `rake build`. Then put that gem on the target system and give it a whirl. Note: you'll need to make sure you've got the proper version of gecode installed on the target node but we're working on dropping that as a core dependency of Berkshelf.

@sethvargo what do you think?

**Comment by sethvargo** [comment]
I think the pseudo-logic should look like this:

```ruby
ENV["HOME"] || ENV["BERKSHELF_PATH"] || ENV["TMPDIR"] || "/tmp"
```

Thoughts?

**Comment by eric-tucker** [comment]
I can look into issuing a PR myself.  However, I've had some terrible luck trying to install the standard gem on Amazon Linux however due to dependency issues.  ChefDK was the only way I could get Berks installed on AWS.  Hopefully I don't encounter these same issues (it has been a while since I last tried the gem).

Does Berkshelf need to work on Windows?  If so, a default of `/tmp` won't work, will it?

Thanks for the positive feedback folks.  It's appreciated.

**Comment by reset** [comment]
@eric-tucker yeah I know that pain, that's why we put it in ChefDK :smile:. We just released a new version of [Solve](https://github.com/berkshelf/solve) which, by default, uses a pure Ruby solver instead of the C based Gecode solver which is the source of the installation pain. I'm planning on updating Berkshelf to use this by default _soon_ which will help these problems.

This doesn't help you right now, though. Installing Berkshelf and Gecode on OSX is easier if that's your dev machine. What OS are you developing on?

Great catch btw - Berkshelf does need to work on Windows. I don't really know what a non-interactive session looks like though (if it's even a thing) in Windows. Is there a temporary directory equivalent in Windows? Maybe we just don't bother with it for now.

**Comment by eric-tucker** [comment]
I'm using Fedora for local work but ultimately any solution has to run on Amazon Linux.

Honestly I'm not sure if Windows would have a non-interactive shell mode.  It is Windows after all and I can't imagine people are doing too much unattended shell work with Windows.  Even if it were a thing, it's not like it's working today without `$HOME` set so the problem would simply be moved from to `/tmp` not existing.  Trading one issue for another, but no better or worse off.

**Comment by reset** [comment]
@eric-tucker it'll work just fine on Amazon Linux once things are pushed into the ChefDK. You just need to get started on your dev machine. You just need to install the right version of gecode and things should be good to go.

I'm not using Windows at all but I do know that things are working well on Windows. I'm not entirely sure if people are setting the HOME env variable themselves, though.

**Comment by JohnMorales** [comment]
since BERKSHELF_PATH seems like it would be more specific than HOME:

 `ENV["BERKSHELF_PATH"] || ENV["HOME"] || ENV["TMPDIR"] || "/tmp"`

**Comment by TheFynx** [comment]
:+1:  On this issue, trying to utilize berks vendor in a cloudformation setup and this makes it impossible.

**Comment by lamont-granquist** [comment]
`ENV["TMPDIR"] || "/tmp"` causing all users without homedirs to write to a well-known path of `/tmp/.berkshelf` in a 1777 directory is likely a security hole and an unacceptable solution.
